src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+                elif future._must_stop:
+                    future._was_stopped = True
+                    raise CancelledError()
+                else:
+                    logging.error("Validation failed: %s", ex)
+                    raise
+            except CancelledError:
+                logging.debug("Validation cancelled.")
+                future._was_stopped = True
+                raise  # No fuss, pass it as-is
+            except Exception:
+                logging.exception("Validation failure.")
+                raise
+            finally:
+                if not self.IsStable():
+                    raise IOError("Validation failed, device not stable.")
+                for ch, num in self.channels.items():
+                    if not self.IsValid(num):
+                        raise IOError("Failed to validate channel %s" % num)
+                self._updatePosition()
+
+    def _on_referenced(self, future):
+        """
+        Callback function for referencing/validation future in __init__.
+        """
+        try:
+            future.result()
+            self.state._set_value(model.ST_RUNNING, force_write=True)
+        except Exception as e:
+            self.state._set_value(e, force_write=True)
+            logging.exception(e)
+
+    def _updatePosition(self):
+        """
+        Updates the position VA.
+        """
+        pos = self._getPosition()
+        self.position._set_value(pos, force_write=True)
+        logging.debug("Updated position to %s.", pos)
+
+    def _getPosition(self):
+        """
+        Getter for .position VA. Requests position from device.
+        returns: dict (str --> float)
+        """
+        if self.state == model.ST_STARTING:
+            logging.warning("Device has not finished initializing, position request will be inaccurate.")
+        pos = {}
+        for name, num in self.channels.items():
+            pos[name] = self.GetValue_f64(num, 0)  # position value is at index 0
+        return pos
+
+    def _createReferenceFuture(self):
+        """
+        returns: (CancellableFuture)
+        """
+        f = CancellableFuture()
+        f._must_stop = False  # cancel of the current future requested
+        f._was_stopped = False  # future was actually stopped
+        f._moving_lock = threading.Lock()  # taken while moving
+        f.task_canceller = self._cancelReference
+        return f
+
+    def _cancelReference(self, future):
+        """
+        Cancels the referencing procedure.
+        future (Future): the future to stop
+        """
+        logging.debug("Cancelling referencing...")
+        future._must_stop = True  # tell the thread taking care of the referencing it's over
+        self.core.SA_SI_Cancel(self._id)
+
+        # Synchronise with the ending of the future
+        with future._moving_lock:
+            if not future._was_stopped:
+                logging.debug("Cancelling failed.")
+            return future._was_stopped
```

Comment:
The main point of the lock is to make sure you call `Cancel()` only when a function is started.
I think that in the cancel() you could write:
``` python
        future._must_stop = True  # tell the thread taking care of the referencing it's over

        # Synchronise with the ending of the future
        with future._moving_lock:
            self.core.SA_SI_Cancel(self._id)
            if not future._was_stopped:
                logging.debug("Cancelling failed.")
            return future._was_stopped
```

The actual move functions would look like:
```python
with future._moving_lock:
    if future._must_stop:
         future._was_stopped = True
    starting_movement_function()

wait_for_event(long timeout)
:
# you can put more start (in locks), followed by waits out of locks
```

Reply:
Is that such a good idea? Doesn't this make it impossible to stop the 5 minute-long referencing procedure?

Reply:
It should still be possible to stop the procedure at almost any time. If that's not the case it's a bug.
The goal of the changes I suggest is only to ensure that it's not possible to cancel and then the procedure starts. (Due to race condition)

Reply:
You're right, I didn't realize that you put `wait_for_event` outside the lock. Then it makes complete sense.